### PR TITLE
Accept qubit maps in transform_qubits

### PR DESCRIPTION
**Pull request by tanujkhattar**
Solves #3611

### CODE REVIEW

File: cirq/circuits/circuit.py [hunk 1]
```diff
@@ -1504,7 +1504,10 @@ def with_device(
         )
 
     def transform_qubits(
-        self, func: Callable[['cirq.Qid'], 'cirq.Qid'], *, new_device: 'cirq.Device' = None
+        self,
+        qubit_map: Union[Dict['cirq.Qid', 'cirq.Qid'], Callable[['cirq.Qid'], 'cirq.Qid']],
```
Comment (mpharrigan) [inline cirq/circuits/circuit.py:1508]: this is technically a breaking change. Consider adding a keyword-only argument `func=None` that emits a deprecation warning and forwards to `qubit_map`

Comment (tanujkhattar) [reply cirq/circuits/circuit.py:1518]: thanks for pointing out. done.

File: cirq/circuits/circuit.py [hunk 2]
```diff
@@ -1503,8 +1503,21 @@ def with_device(
             device=new_device,
         )
 
+    @deprecated_parameter(
+        deadline='v0.10.0',
```
Comment (balopat) [inline cirq/circuits/circuit.py:1507]: Should be 
```suggestion
        deadline='v0.11.0',
```
We'll be releasing v0.10 soon - that's the current dev version

File: cirq/ops/moment.py [hunk 3]
```diff
@@ -279,20 +279,31 @@ def __repr__(self) -> str:
     def __str__(self) -> str:
         return self.to_text_diagram()
 
+    @deprecated_parameter(
+        deadline='v0.10.0',
```
Comment (balopat) [inline cirq/ops/moment.py:283]:
```suggestion
        deadline='v0.11.0',
```

File: cirq/ops/raw_types.py [hunk 4]
```diff
@@ -445,18 +446,40 @@ def with_tags(self, *new_tags: Hashable) -> 'cirq.TaggedOperation':
         """
         return TaggedOperation(self, *new_tags)
 
-    def transform_qubits(self: TSelf, func: Callable[['cirq.Qid'], 'cirq.Qid']) -> TSelf:
+    @deprecated_parameter(
+        deadline='v0.10.0',
```
Comment (balopat) [inline cirq/ops/raw_types.py:450]:
```suggestion
        deadline='v0.11.0',
```

File: cirq/ops/moment_test.py [hunk 5]
```diff
@@ -579,3 +579,16 @@ def test_commutes():
     assert not cirq.commutes(moment, cirq.H(a))
     assert not cirq.commutes(moment, cirq.H(b))
     assert not cirq.commutes(moment, cirq.X(c))
+
+
```
Comment (balopat) [inline cirq/ops/moment_test.py:583]: we should add a test method that tests the deprecation logic as well, ensures the right warning message is logged.

**Review by balopat — CHANGES_REQUESTED** [review]
Thanks for the fix! The deprecation logic needs some more work!

File: cirq/circuits/circuit.py (hunk 2 above)
Comment (tanujkhattar) [reply cirq/circuits/circuit.py]: done

File: cirq/ops/raw_types.py (hunk 4 above)
Comment (tanujkhattar) [reply cirq/ops/raw_types.py]: done

File: cirq/ops/moment.py (hunk 3 above)
Comment (tanujkhattar) [reply cirq/ops/moment.py]: done

File: cirq/ops/moment_test.py (hunk 5 above)
Comment (tanujkhattar) [reply cirq/ops/moment_test.py:583]: added a test for the deprecation logic.

**Review by balopat — APPROVED** [review]

**Pull request merged**